Fix 3541 - remove "admin" permission from shop manager role

Akarshit opened the pull request:
### Fix for #3541 
### To test
- Start the app on this branch
- Notice the migrations will run in your terminal. Something like(Running up() on version 19)
- Invite a user to Shop Manager group
- Remove `email` permission from Shop Manager role
- Login as the invited user, you shouldn’t be able to see the email icon in the sidebar.
- Now invite a second user to the Shop Manager role.
- Remove `i18n` permission from the Shop Manager group.
- Login as the invited user in incognito mode.
- Observe that i18n menu is not available to you.

pmn4 commented:
assuming I have my head straight, there are:

- Marketplace Admins
- Shop Owners
- Shop Managers
- Customers
- Guests

and your change affects Shop Managers, right?
if so, lgtm!

zenweasel commented:
There's a little bit of a discrepancy between this PR and the title of the issue that this is supposed to fix and I want to make sure this PR actually closes it. The title seems to imply that all users are not always seeing the icons based on their permissions, but the fix for this is just adjusting the permissions of one particular group. So am I to assume that except for this group, all users were seeing the appropriate icons based on their permissions @impactmass?

Akarshit commented:
@pmn4 Yes the change only affects Shop Mangers.

impactmass commented:
I'm currently out sick, and not able to review this at the moment.    Yes,
I do agree @zenweasel 's point, and will consider that when I review
(hopefully soon). thanks

On Fri, Jan 19, 2018 at 6:13 AM, Akarshit Wal <notifications@github.com>
wrote:

> @pmn4 <https://github.com/pmn4> Yes the change only affects Shop Mangers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/reactioncommerce/reaction/pull/3505#issuecomment-358868734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE_LCXKpn63_6o2phXMAlAUu9eSjnaoWks5tMCR_gaJpZM4RiimF>
> .
>



-- 
Seun Martins
@impactmass <http://twitter.com/impactmass>
0806-242-2776

impactmass commented:
Hi @Akarshit, thanks for digging into this issue. I went over the ticket again + the code here. Some conclusions:

- The title and expected behavior section of the ticket #3091 described the issue as though it happens to all user. I tested again and I can see that it only happens with users in "Shop Manager" group.  
  - I created a new group, added a user to it, then toggled individual permissions from the group, while confirming that they were being removed from the icons list of the user.
  - So, @zenweasel you are correct by saying: `except for this group, all users were seeing the appropriate icons based on their permissions`

- Recommendation: Close the ticket (#3091). Make another ticket that details the specific problem with the "Shop Manager" group (i.e the problem being caused by including "admin" in its permissions). I recommend a new ticket because there needs to be discussion/decision about keeping/removing the "admin" role.

- I also found a related (but different) issue, where a user with dashboard access can be able to access the dashboard setting panel (not the icon) for a particular package even without having the needed permission(s). I'll write a ticket for this too.

Akarshit commented:
@impactmass Ok will do.

mikemurray commented on imports/plugins/core/versions/server/migrations/5_update_defaultRoles_to_groups.js:
```diff
@@ -57,7 +57,7 @@ Migrations.add({
       ownerRoles = _.uniq(ownerRoles);
 
-      const shopManagerRoles = ownerRoles.filter(role => role !== "owner");
+      const shopManagerRoles = ownerRoles.filter(role => role !== "owner" && role !== "admin");
```
This is modifying an existing migration means it will not run for existing shops. 

Is this change critical enough that this migration must be run to adjust the admin roles for an existing shop? If so, then there will need to be a new migration that removes the admin role from the shop managers group.

Akarshit replied:
@mikemurray In my opinion yes, we should write a migration to adjust the admin roles for existing shops.
@impactmass What do you think about this?

impactmass replied:
@Akarshit I agree it should be a separate migration to adjust it.

impactmass commented:
@Akarshit re: last comment on the ticket about potential side effects that can occur because of this change, I'll hold off review till that is looked into.

Akarshit commented:
@impactmass I can't think of any problem, because we are giving all the individual permissions to the group. Although I don't know enough to be 100% sure.
@spencern @zenweasel Does any problem come to your mind? Or is there some way to check?

spencern commented:
@Akarshit I can't think of anything specifically. As long as we're granting all other permissions by default, it should work out, but there may be places where we're checking specifically against "admin/owner" permissions that we'd need to make note of and update to check a specific permission instead.

Akarshit commented:
I have added the migration to remove `admin` role from `Shop Manager` group but I also changed the 5th migration so that anyone who decides to migrate to <= 18 version also gets the correct roles. 
Also updated the test instruction.

P.S I am away for a week so if something needs to be pushed for this feel free to do so, or I will come back and fix it.

zenweasel commented:
This looks good to me and solves the problem outlined in the ticket
